### PR TITLE
Render all records allowed by the authorization scope

### PR DESCRIPTION
**Pull request by pablobm**
Working with Pundit, I was seeing this index page:

Screenshot from 2019-12-13 22-28-37

That was supposed to be 4 records, but they were showing as empty rows. Instead I was expecting to see the following:

Screenshot from 2019-12-13 22-29-10

The issue was that the Pundit scope for the model was returning those records, but the per-resource policy was returning `false` for `show?` on each one of them.

This is a bit of an edge case: if the user can't see the resources, surely the scope shouldn't be returning them. Still I can think of a use case where some users can see them listed (with limited information) but can't access the show page (with additional information). In any case, Pundit allows for this, and I think this behaviour is a bug.

The fix itself is just in `app/views/administrate/application/_collection.html.erb`. Apart from that, this PR is mostly me figuring out a way to test this in a disposable way that doesn't affect other specs.

"Mostly", because there's one more thing. While working with this, I came across the `tabindex` settings on the index page. As they stand at the moment, each row is tabbable, as well as each individual cell inside each row. I understand this is because these are links, but it strikes me as very annoying to navigate. On the example above, you need to tab 5 times per row, and you get the same link for all 5 positions. I don't think this is the way `tabindex` is supposed to be used.

The reason I made this other change is because it's related: when applying the authorization fix, I had to also make rows tabbable only if they had a link. While changing this, I noticed the other behaviour.

Can someone who knows more about accessibility than me confirm all this?

### CODE REVIEW

File: app/assets/stylesheets/administrate/base/_tables.scss [hunk 1]
```diff
@@ -21,6 +21,9 @@ tr {
 tbody tr {
   &:hover {
     background-color: $base-background-color;
+  }
+
+  &[role=link] {
```
[the remark this answers is not in the document]
Comment (pablobm) [reply app/assets/stylesheets/administrate/base/_tables.scss]: Hm, I think this one is legitimate. Opinions?

**Comment by tabfugnic** [comment]
@pablobm Is this just to prove that Pundit works? Or is Pundit a fundamental requirement of Administrate. If it's not I feel the most important thing is that we don't make it contingent and the code doesn't have to be too specific to support this.

Other than that I'm pretty stoked about this change.

File: app/assets/stylesheets/administrate/base/_tables.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (tabfugnic) [reply app/assets/stylesheets/administrate/base/_tables.scss]: Does this need to be specific to the `tbody tr`? Can this be more general, I think this is what it's complaining about. 

https://stylelint.io/user-guide/rules/selector-no-qualifying-type

[the remark this answers is not in the document]
Comment (pablobm) [reply app/assets/stylesheets/administrate/base/_tables.scss]: So my thinking is that this style covers the specific case where a `table tbody tr` is enabled as a link. This is done with JS at https://github.com/thoughtbot/administrate/blob/669006ae13cdd1067665869adaf50d2817ce6af7/app/assets/javascripts/administrate/components/table.js

On second thoughts, maybe it's ok to move it to apply to any descendant of `tbody`, without qualifying the `tr`. After all, if it's role is a link, it makes sense that the cursor is a pointer regardless of where specifically it is.

**Comment by pablobm** [comment]
As I see it, there are two things at play here:

1. If a record is in the list, it should render fully, not as an empty row.
2. Administrate does explicitly support Pundit, as documented at http://administrate-prototype.herokuapp.com/authorization.

On the first point, I think that even if we were going to use `valid_action?` instead of `show_action?`, it should be in the right place so that this behaviour doesn't occur. Note that `valid_action?` is not about authorization, but instead it's a piece of magic to tell us if the "show" action (in this specific example) exists and therefore if a link makes sense here. If we were only listing but not showing the records, it would be legitimate to render them properly.

On the second point, I use `show_action?` because it's a hook explicitly provided by Administrate to be overriden by authorization mechanisms, such as the bundled `Adminstrate::Punditize`. There are very few and limited examples of this in the specs, so I thought of introducing this new feature spec to provide a way of start testing this better.

Finally, there's the issue of `tabindex`. To be honest, I think I'm cheating a bit by bundling it in here. After a bit of introspection, I realise that the main reason I put it here is because, if I created a separate PR for it, it would have to wait for this one to merge as it would conflict. I'm happy to extract it; it may provide for a better place to discuss it, and to isolate the change if anyone looks for it in commits in the future.

**Comment by pablobm** [comment]
I'm going to split this PR into two. I have started by creating https://github.com/thoughtbot/administrate/pull/1557, which deals with the tabindex issue. Then I'll create another PR dealing with the Pundit issue, and will close this one.

**Comment by pablobm** [comment]
Superseded by https://github.com/thoughtbot/administrate/pull/1596 and https://github.com/thoughtbot/administrate/pull/1557

**Pull request closed without merging**